### ISSUE TITLE
One node instance is always "Out of Service"

### ISSUE THREAD

**Issue by demisx** [kubernetes/kops#8759]
**1. What `kops` version are you running? 
`Version 1.16.0`

**2. What Kubernetes version are you running? `kubectl version` will print the
 version if a cluster is running or provide the Kubernetes version specified as
 a `kops` flag.**
```
Client Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.4", GitCommit:"xxxxxxxxxxxx", GitTreeState:"clean", BuildDate:"2020-03-12T23:41:24Z", GoVersion:"go1.14", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.7", GitCommit:"xxxxxxxxxxx", GitTreeState:"clean", BuildDate:"2020-02-11T19:24:46Z", GoVersion:"go1.13.6", Compiler:"gc", Platform:"linux/amd64"}
```

**3. What cloud provider are you using?**
`AWS`

**4. What commands did you run?  What is the simplest way to reproduce this issue?**
Create new cluster from scratch

**5. What happened after the commands executed?**
One of the node instances is marked by ELB as "OutOfService".  However, all clusters are successfully validated with `kops validate cluster`. We have 3 different clusters running and they all suffer from this issue.

```sh
i-04a7295xxxxxxxxx | nodes-us-west-2a.prod.dimaslist.org | us-west-2a | InService
i-0051712xxxxxxxxx | nodes-us-west-2b.prod.dimaslist.org | us-west-2b | InService
i-0efb2042xxxxxxxx | nodes-us-west-2d.prod.dimaslist.org | us-west-2d | OutOfService
```

**6. What did you expect to happen?**
All node instances will be marked by ELB as "In Service"

**7. Please provide your cluster manifest. Execute
  `kops get --name my.example.com -o yaml` to display your cluster manifest.
  You may want to remove your cluster name and other sensitive information.**
```yaml
apiVersion: kops.k8s.io/v1alpha2
kind: Cluster
metadata:
  creationTimestamp: "2019-10-22T17:23:28Z"
  generation: 8
  name: prod.example.org
spec:
  api:
    loadBalancer:
      type: Public
  authorization:
    rbac: {}
  channel: stable
  cloudProvider: aws
  configBase: s3://kops-example-org-prod/prod.example.org
  dnsZone: Z1KK30M1ET94CJ
  etcdClusters:
  - cpuRequest: 200m
    etcdMembers:
    - encryptedVolume: true
      instanceGroup: master-us-west-2a
      name: a
    - encryptedVolume: true
      instanceGroup: master-us-west-2b
      name: b
    - encryptedVolume: true
      instanceGroup: master-us-west-2d
      name: d
    memoryRequest: 100Mi
    name: main
    version: 3.2.24
  - cpuRequest: 100m
    etcdMembers:
    - encryptedVolume: true
      instanceGroup: master-us-west-2a
      name: a
    - encryptedVolume: true
      instanceGroup: master-us-west-2b
      name: b
    - encryptedVolume: true
      instanceGroup: master-us-west-2d
      name: d
    memoryRequest: 100Mi
    name: events
    version: 3.2.24
  iam:
    allowContainerRegistry: true
    legacy: false
  kubeAPIServer:
    featureGates:
      TTLAfterFinished: "true"
  kubeControllerManager:
    featureGates:
      TTLAfterFinished: "true"
  kubeProxy:
    metricsBindAddress: 0.0.0.0
  kubelet:
    anonymousAuth: false
    featureGates:
      TTLAfterFinished: "true"
  kubernetesApiAccess:
  - 0.0.0.0/0
  kubernetesVersion: 1.16.7
  masterInternalName: api.internal.prod.example.org
  masterPublicName: api.prod.example.org
  networkCIDR: 172.20.0.0/16
  networking:
    calico:
      majorVersion: v3
  nonMasqueradeCIDR: 100.64.0.0/10
  sshAccess:
  - 0.0.0.0/0
  subnets:
  - cidr: 172.20.32.0/19
    name: us-west-2a
    type: Private
    zone: us-west-2a
  - cidr: 172.20.64.0/19
    name: us-west-2b
    type: Private
    zone: us-west-2b
  - cidr: 172.20.96.0/19
    name: us-west-2d
    type: Private
    zone: us-west-2d
  - cidr: 172.20.0.0/22
    name: utility-us-west-2a
    type: Utility
    zone: us-west-2a
  - cidr: 172.20.4.0/22
    name: utility-us-west-2b
    type: Utility
    zone: us-west-2b
  - cidr: 172.20.8.0/22
    name: utility-us-west-2d
    type: Utility
    zone: us-west-2d
  topology:
    dns:
      type: Public
    masters: private
    nodes: private

---

apiVersion: kops.k8s.io/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: "2020-03-15T21:02:54Z"
  labels:
    kops.k8s.io/cluster: prod.example.org
  name: bastions
spec:
  image: kope.io/k8s-1.16-debian-stretch-amd64-hvm-ebs-2020-01-17
  machineType: t2.micro
  maxSize: 1
  minSize: 1
  nodeLabels:
    kops.k8s.io/instancegroup: bastions
  role: Bastion
  subnets:
  - utility-us-west-2a

---

apiVersion: kops.k8s.io/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: "2019-10-22T17:23:28Z"
  generation: 3
  labels:
    kops.k8s.io/cluster: prod.example.org
  name: master-us-west-2a
spec:
  image: kope.io/k8s-1.16-debian-stretch-amd64-hvm-ebs-2020-01-17
  machineType: t3a.medium
  maxSize: 1
  minSize: 1
  nodeLabels:
    kops.k8s.io/instancegroup: master-us-west-2a
  role: Master
  subnets:
  - us-west-2a

---

apiVersion: kops.k8s.io/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: "2019-10-22T17:23:28Z"
  generation: 3
  labels:
    kops.k8s.io/cluster: prod.example.org
  name: master-us-west-2b
spec:
  image: kope.io/k8s-1.16-debian-stretch-amd64-hvm-ebs-2020-01-17
  machineType: t3a.medium
  maxSize: 1
  minSize: 1
  nodeLabels:
    kops.k8s.io/instancegroup: master-us-west-2b
  role: Master
  subnets:
  - us-west-2b

---

apiVersion: kops.k8s.io/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: "2019-10-22T17:23:28Z"
  generation: 3
  labels:
    kops.k8s.io/cluster: prod.example.org
  name: master-us-west-2d
spec:
  image: kope.io/k8s-1.16-debian-stretch-amd64-hvm-ebs-2020-01-17
  machineType: t3a.medium
  maxSize: 1
  minSize: 1
  nodeLabels:
    kops.k8s.io/instancegroup: master-us-west-2d
  role: Master
  subnets:
  - us-west-2d

---

apiVersion: kops.k8s.io/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: "2019-10-22T17:26:50Z"
  generation: 3
  labels:
    kops.k8s.io/cluster: prod.example.org
  name: nodes-us-west-2a
spec:
  image: kope.io/k8s-1.16-debian-stretch-amd64-hvm-ebs-2020-01-17
  machineType: t3a.large
  maxPrice: "0.05"
  maxSize: 5
  minSize: 1
  mixedInstancesPolicy:
    instances:
    - t3a.large
    - t3.large
    - r5a.large
    - m5.large
    onDemandAboveBase: 0
    spotInstancePools: 4
  nodeLabels:
    kops.k8s.io/instancegroup: nodes-us-west-2a
  role: Node
  subnets:
  - us-west-2a

---

apiVersion: kops.k8s.io/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: "2019-10-22T17:27:42Z"
  generation: 3
  labels:
    kops.k8s.io/cluster: prod.example.org
  name: nodes-us-west-2b
spec:
  image: kope.io/k8s-1.16-debian-stretch-amd64-hvm-ebs-2020-01-17
  machineType: t3a.large
  maxPrice: "0.05"
  maxSize: 5
  minSize: 1
  mixedInstancesPolicy:
    instances:
    - t3a.large
    - t3.large
    - r5a.large
    - m5.large
    onDemandAboveBase: 0
    spotInstancePools: 4
  nodeLabels:
    kops.k8s.io/instancegroup: nodes-us-west-2b
  role: Node
  subnets:
  - us-west-2b

---

apiVersion: kops.k8s.io/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: "2019-10-22T17:28:34Z"
  generation: 3
  labels:
    kops.k8s.io/cluster: prod.example.org
  name: nodes-us-west-2d
spec:
  image: kope.io/k8s-1.16-debian-stretch-amd64-hvm-ebs-2020-01-17
  machineType: t3a.large
  maxPrice: "0.05"
  maxSize: 5
  minSize: 1
  mixedInstancesPolicy:
    instances:
    - t3a.large
    - t3.large
    - r5a.large
    - m5.large
    onDemandAboveBase: 0
    spotInstancePools: 4
  nodeLabels:
    kops.k8s.io/instancegroup: nodes-us-west-2d
  role: Node
  subnets:
  - us-west-2d
```

**9. Anything else do we need to know?**
The ELB health check is:
```
Ping Target | HTTP:32502/healthz
Timeout | 5 seconds
Interval | 10 seconds
Unhealthy threshold | 6
Healthy threshold | 2
```
This is what I am getting back from each node. The "OutOf service" one returns "503 Service Unavailable".  Any idea what may cause one of the nodes to return 503?

```
$ curl -I 172.20.53.122:32502/healthz
HTTP/1.1 200 OK
Content-Type: application/json
Date: Sun, 15 Mar 2020 21:45:26 GMT
Content-Length: 105

$ curl -I 172.20.94.71:32502/healthz
HTTP/1.1 200 OK
Content-Type: application/json
Date: Sun, 15 Mar 2020 21:47:19 GMT
Content-Length: 105

$ curl -I 172.20.127.105:32502/healthz
HTTP/1.1 503 Service Unavailable
Content-Type: application/json
Date: Sun, 15 Mar 2020 21:47:47 GMT
Content-Length: 105
```

The `/healthz` endpoint doesn't exist on the "out of service" node:
```sh
$ curl localhost:32502/healthz
{
	"service": {
		"namespace": "default",
		"name": "nginx-ingress-controller"
	},
	"localEndpoints": 0 # <-- the "in service" nodes show this value as 1
}
```

Changing the ELB health check from HTTP to TCP immediately puts all nodes in service, but this does not mean that all 3 are properly working, of course. In fact, the previously out of service node starts causing issues if put into service. So, something is wrong with it.

**Comment by demisx** [comment]
If anyone could please give some tips where to look for the problem I'd greatly appreciate it. Obviously, there is a web server that fails to respond to `/healthz` endpoint. But what type of server and where is it? I can't seem to find the log that would shed some lights on it, either. Thank you.

**Comment by nirupbbnk** [comment]
We are also facing same issue. Nodes are healthy in  AWS console, But one node will be randomly in Notready state. After some times this node also become ready state. Kube version v1.16.7 , kops v1.16.0

**Comment by demisx** [comment]
In my case, the "OutOfService" node is not working properly indeed, but I am not clear where to start troubleshooting from because this node is bootstrapped by the Kops itself.  I've check all all logs I could think of but didn't see any red flags. Any suggestions are highly appreciated.

**Comment by strangeminds** [comment]
We've also seen this issue once.  One day we started to see "504 Gateway Timeout" errors in our Application Load Balancer logs.  We traced the issue to one new node that was built using auto-scaling.  Checking the ALB target group and would see the one new node "OutOfService" and a few minutes later be fine, then back to "OutOfService".  When we would describe the node, we saw under event conditions that the node was unhealthy.

kubectl describe node ip-XXX-XXX-XXX-XXX.ec2.internal 

Our workaround was to drain the one new node to remove it from the cluster.  In our case, we figured it was just a bad EC2 instance from AWS.  Since then we've had auto-scaling work just fine and the issue has not popped back up.

**Comment by demisx** [comment]
I realized that I have two load balancers:

1. Provisioned by Kops - no issues , all nodes are in service
1. Provisioned by Nginx Ingress - one node is out of service 

The problem with the 2nd ELB appears to be with the scheduler placing 1 nginx controller pod on one node and the remaining 2 nginx controller pods on another node, thus the 3rd node has no nginx controller pod running at all. Therefore, there is nothing on that node to respond to the `/healthz` request and the ELB correctly marks this node as out of service. I just have to figure out how to force the k8s scheduler to always place one nginx controller pod per one individual node. 

Closing this issue as it appears not related to Kops functionality.
